File: node/src/main/java/dev.httpmarco.polocloud.node/commands/specific/GroupArgument.java
package dev.httpmarco.polocloud.node.commands.specific;

import dev.httpmarco.polocloud.api.Named;
import dev.httpmarco.polocloud.api.groups.ClusterGroup;
import dev.httpmarco.polocloud.node.Node;
import dev.httpmarco.polocloud.node.commands.CommandArgument;
import dev.httpmarco.polocloud.api.groups.ClusterGroupProvider;
import dev.httpmarco.polocloud.node.commands.CommandContext;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Unmodifiable;

import java.util.List;
import java.util.Objects;

public final class GroupArgument extends CommandArgument<ClusterGroup> {


    public GroupArgument(String key) {
        super(key);
    }

    @Override
    public boolean predication(@NotNull String rawInput) {
        return Node.instance().groupProvider().exists(rawInput);
    }

    @Contract(pure = true)
    @Override
    public @NotNull String wrongReason() {
        return "The Argument " + key() + " is not a registered cluster group!";
    }

    @Override
    public @NotNull @Unmodifiable List<String> defaultArgs(CommandContext context) {
        return Node.instance().groupProvider().groups().stream().map(Named::name).toList();
    }

    @Contract("_ -> new")
    @Override
    public @NotNull ClusterGroup buildResult(String input) {
        return Objects.requireNonNull(Node.instance().groupProvider().find(input));
    }
}

File: agent/src/main/java/dev/httpmarco/polocloud/agent/runtime/local/terminal/commands/type/GroupArgument.kt
package dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.type

import dev.httpmarco.polocloud.agent.Agent
import dev.httpmarco.polocloud.agent.groups.Group
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandArgument
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandContext
import dev.httpmarco.polocloud.agent.runtime.local.terminal.commands.CommandSyntax

class GroupArgument(key: String = "group") : CommandArgument<Group>(key) {

    override fun buildResult(input: String, context: CommandContext): Group {
        return Agent.instance.runtime.groupStorage().item(input)!!
    }

    override fun defaultArgs(context: CommandContext): MutableList<String> {
        return Agent.instance.runtime.groupStorage().items().map { it.data.name }.toMutableList()
    }

    override fun predication(rawInput: String): Boolean {
        return Agent.instance.runtime.groupStorage().present(rawInput)
    }
}